### PR TITLE
Add crypto contribution flow elements

### PR DESCRIPTION
**Pull request by SudharakaP**
⚠️ Should only be merged after completing, https://github.com/opencollective/opencollective/issues/4456

Resolve https://github.com/opencollective/opencollective/issues/4484

basic-contribution-flow

### CODE REVIEW

**Comment by SudharakaP** [comment]
Here I've added the basic support for the crypto flow. Now I think I am gonna start working on the integration with the giving block part. One thing to note is that I didn't yet add the conversion of crypto on the fly (using probably [fixer.io](https://fixer.io/) 🤔 ) so that we have an instant conversion on the front-end. 

image

On this one I have a couple of questions; 

- Are we good to use [fixer.io](https://fixer.io/) for this? 🤔 

- Maybe we'll do this at the end after we work on the integration of giving block etc. My thinking is that this isn't strictly necessary for the MVP? 🤔 

cc: @kewitz @opencollective/core-developers

**Comment by znarf** [comment]
> Are we good to use fixer.io for this? 🤔

I don't know if fixer support this. Also fixer is not configured yet on the frontend, and I don't think it's suitable on the client side anyway.

> Maybe we'll do this at the end after we work on the integration

Yes, definitely.

If it was simple, I would not mind. But I see it as a distraction for now.

**Comment by SudharakaP** [comment]
> > Are we good to use fixer.io for this? thinking
> 
> I don't know if fixer support this. Also fixer is not configured yet on the frontend, and I don't think it's suitable on the client side anyway.
> 
> > Maybe we'll do this at the end after we work on the integration
> 
> Yes, definitely.
> 
> If it was simple, I would not mind. But I see it as a distraction for now.

Thanks for the confirmation. I think it needs a subscription if we plan to use it long term. Maybe there's better alternatives, so let me come back to it later then. 🤔

**Comment by znarf** [comment]
> I think it needs a subscription if we plan to use it long term.

We have a subscription.

**Comment by SudharakaP** [comment]
> > I think it needs a subscription if we plan to use it long term.
> 
> We have a subscription.

Oh that's nice, could you share it with me or add it to `1password`, so maybe I can take a look creating and api key maybe and trying that out as well later. 🤔

File: lang/pt-BR.json
Comment (kewitz) [inline lang/pt-BR.json:1932]: I think we shouldn't discard this because of a typo fix.

File: lang/uk.json
Comment (kewitz) [inline lang/uk.json:1942]: For some reason, this file is still outdated.

File: components/collective-page/sections/Contribute.js
Comment (kewitz) [inline components/collective-page/sections/Contribute.js]: This means the default behavior is disabled, is this correct?

File: components/contribution-flow/CryptoStepDetails.js
Comment (kewitz) [inline components/contribution-flow/CryptoStepDetails.js]: I think you can drag this to a constants file.

File: components/contribution-flow/index.js
Comment (kewitz) [inline components/contribution-flow/index.js:657]: I think it would be simpler if you declared a variable like `isCrypto` in this page and dealt with that from this level-down.

File: components/collective-page/sections/Contribute.js
Comment (SudharakaP) [reply components/collective-page/sections/Contribute.js]: @kewitz : Yes, that's the idea. They need to explicitly enable in order to display the crypto card. 🤔 🙂

File: components/contribution-flow/CryptoStepDetails.js
Comment (SudharakaP) [reply components/contribution-flow/CryptoStepDetails.js]: Done. 👍🏽

File: components/contribution-flow/index.js
Comment (SudharakaP) [reply components/contribution-flow/index.js]: Done. 👍🏽

File: lang/uk.json
Comment (SudharakaP) [reply lang/uk.json:1941]: I've updated the lang files, now it should be good. Now sure how this happened, maybe I didn't update it the first time. 🤔

File: lang/pt-BR.json
Comment (SudharakaP) [reply lang/pt-BR.json:1931]: I've updated the lang files, now it should be good. 👍🏽

File: components/contribution-flow/ContributionFlowHeader.js
Comment (znarf) [inline components/contribution-flow/ContributionFlowHeader.js:61]: I don't see this as necessary. Why was this added?

File: components/contribution-flow/ContributionFlowStepContainer.js
Comment (znarf) [inline components/contribution-flow/ContributionFlowStepContainer.js:200]: Why is this necessary?

File: components/contribution-flow/ContributionFlowStepsProgress.js
Comment (znarf) [inline components/contribution-flow/ContributionFlowStepsProgress.js:42]: What's blocking to use FormattedMoneyAmount?

File: components/contribution-flow/ContributionSummary.js
Comment (znarf) [inline components/contribution-flow/ContributionSummary.js:89]: Why not handle this in FormattedMoneyAmount?

File: components/contribution-flow/index.js
Comment (znarf) [inline components/contribution-flow/index.js:147]: What's this?

Why `this.props` vs `props`?

File: components/contribute-cards/ContributeCrypto.js
Comment (znarf) [inline components/contribute-cards/ContributeCrypto.js:23]: Not sure about this. Rationale?

File: components/contribution-flow/ContributionBlocker.js
Comment (znarf) [inline components/contribution-flow/ContributionBlocker.js:75]: Doesn't take into account that the default is true.

File: components/contribution-flow/ContributionFlowStepsProgress.js
Comment (znarf) [inline components/contribution-flow/ContributionFlowStepsProgress.js:80]: Why is this necessary? It seems fine if stepPayment?.paymentMethod is null.

File: next.config.js
Comment (znarf) [inline next.config.js:33]: Not used.

File: components/contribution-flow/ContributionFlowHeader.js
Comment (SudharakaP) [reply components/contribution-flow/ContributionFlowHeader.js:61]: @znarf : I wanted to hide the contributor avatars. If we do want to show the contributors avatars we need to differentiate between custom contributors vs crypto contributors. We don't have that support yet. My thinking is not to do this on the MVP. Let me know if you think otherwise. 🤔

File: components/contribution-flow/ContributionFlowStepContainer.js
Comment (SudharakaP) [reply components/contribution-flow/ContributionFlowStepContainer.js:200]: @znarf : Note that according to the design we don't show a section header for the checkout/payment screen. We just show the QR code. Now I did the QR code addition in a separate PR, but wanted to make the other parts of the contribution flow as close as possible to the design.

https://www.figma.com/file/ZQBMWhnGGtRWeIZknFW1eP/%5BOC-Design%5D-Production-Ready-%E2%9C%85?node-id=13426%3A207907

File: components/contribution-flow/ContributionFlowStepsProgress.js
Comment (SudharakaP) [reply components/contribution-flow/ContributionFlowStepsProgress.js:42]: @znarf : I wanted to keep things simple by not messing up the `FormattedMoneyAmount`. Note that `FormattedMoneyAmount` uses `Currency`, and that assumes we are passing the value in cents; i.e, I don't think we thought about crypto currencies when writing these components. Also note that the currency symbol library that we use `currency-symbol-map` doesn't have all the crypto currency symbols that we need; https://github.com/bengourley/currency-symbol-map/blob/master/map.js. For example I don't see `STORJ` on the list.

Comment (SudharakaP) [reply components/contribution-flow/ContributionFlowStepsProgress.js:42]: But please feel free to let me know if you have suggestions or better alternatives. 🤔

File: components/contribution-flow/ContributionSummary.js
Comment (SudharakaP) [reply components/contribution-flow/ContributionSummary.js:89]: Same reason as https://github.com/opencollective/opencollective-frontend/pull/6689#discussion_r682151172

File: components/contribute-cards/ContributeCrypto.js
Comment (SudharakaP) [reply components/contribute-cards/ContributeCrypto.js:23]: @znarf : My thinking here was that we have `donate` route; and `crypto` to me falls under that category; it's a special kinda tier just like the `Custom Contribution`. It has a special card similar to the `Custom Contribution` card. So I thought it makes sense to be consistent with the routes and make it similar to the `donate` route. The `contribute` routes seems only for specific contribution tiers that the user creates; not the predefined ones such as the `Custom Contribution` or `Crypto` in this case. Those were my thoughts; but I am open to changing this if the team thinks otherwise. 😺

File: components/contribution-flow/ContributionFlowStepsProgress.js
Comment (SudharakaP) [reply components/contribution-flow/ContributionFlowStepsProgress.js:80]: @znarf : If this condition wasn't there we will see the payment method below the section step; 

image

We don't want this for crypto as per the design. As there's no payment method to choose from such as visa or bank account. This is why I added this one.

File: next.config.js
Comment (SudharakaP) [reply next.config.js:33]: Ops, my mistake. I somehow added this again. I've corrected it. 👍🏽

File: components/contribution-flow/ContributionBlocker.js
Comment (SudharakaP) [reply components/contribution-flow/ContributionBlocker.js]: This condition should not be there actually; I wrongly put it there. The contribution should not be blocked for not having cyrpto contributions enabled. Corrected it.

File: components/contribution-flow/index.js
Comment (SudharakaP) [reply components/contribution-flow/index.js]: Ops, my mistake. 😢 Just corrected it.

File: components/contribution-flow/ContributionFlowStepsProgress.js
Comment (znarf) [reply components/contribution-flow/ContributionFlowStepsProgress.js:80]: If `stepPayment?.paymentMethod` is null or undefined, that should not be necessary.

If it is defined, what's its value?

File: components/contribute-cards/ContributeCrypto.js
Comment (znarf) [reply components/contribute-cards/ContributeCrypto.js:23]: `donate` and `contribute` are actions, `crypto` is not. So it feels not proper as URL architecture.

File: components/contribution-flow/ContributionFlowHeader.js
Comment (znarf) [reply components/contribution-flow/ContributionFlowHeader.js:61]: If contributors is empty that's not necessary.

File: components/contribution-flow/ContributionFlowStepsProgress.js
Comment (znarf) [reply components/contribution-flow/ContributionFlowStepsProgress.js:42]: It's not that simpler because you're repeating it multiple times. Just make sure `FormattedMoneyAmount` or `Currency` return what you need when necessary. Detect early on that it's a crypto currency and return what's needed.

File: components/contribution-flow/ContributionSummary.js
Comment (znarf) [reply components/contribution-flow/ContributionSummary.js:89]: Please change.

File: components/contribution-flow/index.js
Comment (znarf) [inline components/contribution-flow/index.js:147]: What is the magic value `5`?

File: components/contribute-cards/ContributeCrypto.js
Comment (SudharakaP) [reply components/contribute-cards/ContributeCrypto.js]: Makes sense, I've made the changed. 👍🏽

File: components/contribution-flow/ContributionSummary.js
Comment (SudharakaP) [reply components/contribution-flow/ContributionSummary.js]: Changed. 👍🏽

File: components/contribution-flow/ContributionFlowStepsProgress.js
Comment (SudharakaP) [reply components/contribution-flow/ContributionFlowStepsProgress.js]: @znarf : Makes sense, I've corrected that to include the logic to format the crypto currencies within `FormattedMoneyAmount`.

File: components/contribution-flow/ContributionFlowHeader.js
Comment (SudharakaP) [reply components/contribution-flow/ContributionFlowHeader.js]: Ah yes, that makes sense. I've corrected that.

File: components/contribution-flow/ContributionFlowStepsProgress.js
Comment (SudharakaP) [reply components/contribution-flow/ContributionFlowStepsProgress.js:80]: Yes, this condition is no longer needed when we add the other parts (the QR code part and stuff). I've removed this at the other PR; https://github.com/opencollective/opencollective-frontend/pull/6725. 👍🏽

File: components/contribution-flow/index.js
Comment (SudharakaP) [reply components/contribution-flow/index.js]: Yes, as discussed I've defaulted this to 0 and fixed the behavior such that when the use first visits this section the input box is auto focused and when the focus out of it the validation takes place. 👍🏽 

validation-on-input

File: components/contribution-flow/ContributionFlowButtons.js
Comment (znarf) [inline components/contribution-flow/ContributionFlowButtons.js:48]: I don't see what justify different words here. I guess it was in the design but still.

File: components/contribution-flow/ContributionFlowStepContainer.js
Comment (znarf) [inline components/contribution-flow/ContributionFlowStepContainer.js:14]: `StepDetailsCrypto` would be better I think.

File: components/contribution-flow/ContributionFlowStepsProgress.js
Comment (znarf) [inline components/contribution-flow/ContributionFlowStepsProgress.js:31]: Why not pass directly as `currency`?

File: components/contribution-flow/ContributionSummary.js
Comment (znarf) [inline components/contribution-flow/ContributionSummary.js:56]: Same as before, why not rely simply on `currency`?

File: components/contribution-flow/index.js
Comment (znarf) [inline components/contribution-flow/index.js:629]: console.log

File: components/contribution-flow/ContributionFlowButtons.js
Comment (SudharakaP) [reply components/contribution-flow/ContributionFlowButtons.js:48]: I am fine with changing this, but I feel the rationale behind this design. My thinking is that the `Payment` header gives the user the idea that they need to choose the payment whereas the checkout page is instantaneous checking out. No payment, it's already checked out. I don't know, it's a slight difference, but the working makes sense to me. They don't do a payment there, they just checkout. Payment is actually done in their wallet side. 😉

File: components/contribution-flow/ContributionFlowStepContainer.js
Comment (SudharakaP) [reply components/contribution-flow/ContributionFlowStepContainer.js]: Done. 👍🏽

File: components/contribution-flow/index.js
Comment (SudharakaP) [reply components/contribution-flow/index.js]: Ops, corrected. 👀

File: components/contribute-cards/ContributeCrypto.js
Comment (SudharakaP) [reply components/contribute-cards/ContributeCrypto.js]: @znarf : Actually thinking about this further I changed this to `/payment/crypto` instead of `/contribute/crypto`. It seems that we will have route conflicts if we use the `contribute` verb. It's also mentioned here; 

https://github.com/opencollective/opencollective-frontend/blob/0247cad01dc7a84034265d7f3ae7d65d1ce6fea1/components/contribution-flow/index.js#L434

Let me know your thoughts or if you have a better suggestion for the naming of the verb.

File: components/contribution-flow/ContributionFlowStepsProgress.js
Comment (SudharakaP) [reply components/contribution-flow/ContributionFlowStepsProgress.js]: I've corrected this. I have a feeling this might be an issue when we create an order maybe, but I'll leave that for the other PR. For this one I've added logic to get the crypto currency based on the flow; 

https://github.com/opencollective/opencollective-frontend/blob/b063aaa834459c4f6a1d168de829a29487dbbf79/components/contribution-flow/index.js#L631

That way we can directly pass the `currency`. 👍🏽

File: components/contribution-flow/ContributionSummary.js
Comment (SudharakaP) [reply components/contribution-flow/ContributionSummary.js]: Corrected as per above; https://github.com/opencollective/opencollective-frontend/pull/6689#discussion_r684481213

File: components/contribute-cards/ContributeCrypto.js
Comment (znarf) [inline components/contribute-cards/ContributeCrypto.js:23]: Why introduce this `payment` suddenly?

File: components/contribution-flow/ContributionFlowButtons.js
Comment (znarf) [reply components/contribution-flow/ContributionFlowButtons.js:48]: What does "Checkout" means actually?

https://dictionary.cambridge.org/dictionary/english/checkout

> the area of a website where you order and pay for goods, or the process of paying for goods

If we take the first part of the definition, without a shopping cart, when you enter the contribution flow, you're already in the "checkout" process. Being the whole process, it's difficult to justify it as a name for a sub-step.  

Looking at the second part of the definition, it's the same as "Payment".

While I'm willing to understand that the payment process with crypto is a bit different than usual, after looking at definitions, I don't understand how introducing this particular new word in the button is helping our users here.

If we want to reuse what's already there:

- "Payment" as in Step Profile
- "Contribute 20 BTC" as in Step Payment method

File: components/contribute-cards/ContributeCrypto.js
Comment (SudharakaP) [reply components/contribute-cards/ContributeCrypto.js:23]: @znarf : Here my thinking was to introduce a verb other than `contribute` as we cannot use `contribute` as explained above; https://github.com/opencollective/opencollective-frontend/pull/6689#discussion_r684470100. I thought that since crypto is a payment method, we add the verb `payment`. Maybe it's not the best; do you any alternative suggestions? 🤔

File: components/contribution-flow/ContributionFlowButtons.js
Comment (SudharakaP) [reply components/contribution-flow/ContributionFlowButtons.js]: @znarf : Makes sense, I've made the changes. Thanks much for the explanation. 😉

Comment (znarf) [reply components/contribution-flow/ContributionFlowButtons.js]: @SudharakaP Thank you for doing the change. I feel a bit sorry to have engaged for such a small change.

But let's keep this in mind has a good guideline, we should avoid introducing new words and new copy if there is not a clear value for the end user. 

This is not final, if we have a good case for a different copy, and help users better understand what they're doing, I'm up for it.

File: components/contribution-flow/index.js
Comment (znarf) [reply components/contribution-flow/index.js]: I don't see zero has a good default value either. Could it perhaps be empty instead?

File: components/contribution-flow/ContributionFlowButtons.js
Comment (SudharakaP) [reply components/contribution-flow/ContributionFlowButtons.js]: Sure no worries. I'll keep this in mind for the future. 👍🏽 😃

File: components/contribute-cards/ContributeCrypto.js
Comment (znarf) [reply components/contribute-cards/ContributeCrypto.js:23]: `payment` is not a verb.

Comment (znarf) [reply components/contribute-cards/ContributeCrypto.js:23]: If `contribute` is not an option, why not use `donate`?

File: components/contribution-flow/index.js
Comment (SudharakaP) [reply components/contribution-flow/index.js]: Yes, it's possible and I've changed this to have an empty default value. 👍🏽

File: components/contribute-cards/ContributeCrypto.js
Comment (SudharakaP) [reply components/contribute-cards/ContributeCrypto.js]: Done. I've changed this to `donate/crypto` instead. I think this should be okay. 🤔

File: rewrites.js
Comment (znarf) [inline rewrites.js:225]: Why do we need the previous one if we have this?

File: rewrites.js
Comment (znarf) [inline rewrites.js:249]: Why do we need the previous one if we have this?

File: components/contribution-flow/index.js
Comment (znarf) [inline components/contribution-flow/index.js:144]: The magical 5 is still there?

Comment (znarf) [reply components/contribution-flow/index.js:144]: I've read that this might be removed in another PR. Why not this one?

**Review by znarf — APPROVED** [review]
Can you look at the latest comments, correct if they make sense and consider this PR approved. We absolutely need to move forward with the project and seems than everything else is now dependent on it.

File: rewrites.js
Comment (SudharakaP) [reply rewrites.js]: You are right, we can combine both of the routes to a single one. I've done that. 👍🏽

File: rewrites.js
Comment (SudharakaP) [reply rewrites.js]: You are right, we can combine both of the routes to a single one. I've done that. 👍🏽

File: components/contribution-flow/index.js
Comment (SudharakaP) [reply components/contribution-flow/index.js:144]: I thought since more updated code are is there we could do it there and this PR is getting a bit long. Maybe I should have put it here. But if it's okay we can revisit this there after this is merged, else I'll do a cherry pick?

Comment (SudharakaP) [reply components/contribution-flow/index.js]: On second thought I've added it to this PR; now it's all good I think. 👍🏽

**Comment by SudharakaP** [comment]
> Can you look at the latest comments, correct if they make sense and consider this PR approved. We absolutely need to move forward with the project and seems than everything else is now dependent on it.

Yes, I've corrected your latest comments, so I think we are good. I am going to wait until the CI is passed and will merge this off today. 🤔 😃

**Pull request merged**
